File: aais-middleware/src/aais_tasks/graph_sync.ts
/**
 * Mythic: Graph ↔ AAIS task bridge
 * Engineering: syncFromGraph / syncToGraph — fail closed; no silent fake
 */
import {
  callGraph,
  graphCreateTodoTask,
  graphListTodoTasks,
  type FetchLike,
} from "../provider_adapters/graph_client.js";
import type { AaisTask } from "./aais_task_model.js";
import type { AaisTaskStore } from "./aais_task_store.js";

export interface GraphSyncResult {
  ok: boolean;
  reasonCode: string;
  imported?: AaisTask[];
  exported?: AaisTask[];
  error?: string;
  needsAuth?: boolean;
}

function mapGraphStatus(s: unknown): AaisTask["status"] {
  const v = String(s || "").toLowerCase();
  if (v === "completed") return "completed";
  if (v === "inprogress" || v === "in_progress") return "inProgress";
  return "notStarted";
}

export async function syncFromGraph(
  store: AaisTaskStore,
  token: string | undefined,
  opts?: { fetchImpl?: FetchLike; listId?: string },
): Promise<GraphSyncResult> {
  if (!token) {
    return {
      ok: false,
      needsAuth: true,
      reasonCode: "GRAPH_SYNC_NEEDS_AUTH",
      error: "Set AAIS_MS_GRAPH_TOKEN or connect Microsoft 365 for syncFromGraph.",
    };
  }
  const listed = await graphListTodoTasks(token, {
    fetchImpl: opts?.fetchImpl,
    listId: opts?.listId,
  });
  if (!listed.ok) {
    return {
      ok: false,
      reasonCode: listed.reasonCode,
      error: listed.error || "Graph list failed",
    };
  }
  const value = (listed.data as { value?: Record<string, unknown>[] })?.value || [];
  const imported: AaisTask[] = [];
  for (const item of value) {
    const graphId = String(item.id || "");
    if (!graphId) continue;
    const existing = store.list().find((t) => t.graphId === graphId);
    if (existing) {
      const updated = store.update(existing.id, {
        title: String(item.title || existing.title),
        status: mapGraphStatus(item.status),
        source: "graph",
      });
      if (updated) imported.push(updated);
      continue;
    }
    imported.push(
      store.create({
        title: String(item.title || "Graph task"),
        status: mapGraphStatus(item.status),
        dueDate: item.dueDateTime
          ? String((item.dueDateTime as { dateTime?: string }).dateTime || "")
          : undefined,
        source: "graph",
        graphId,
      }),
    );
  }
  return { ok: true, reasonCode: "GRAPH_SYNC_FROM_OK", imported };
}

export async function syncToGraph(
  store: AaisTaskStore,
  token: string | undefined,
  taskId: string,
  opts?: { fetchImpl?: FetchLike; listId?: string },
): Promise<GraphSyncResult> {
  if (!token) {
    return {
      ok: false,
      needsAuth: true,
      reasonCode: "GRAPH_SYNC_NEEDS_AUTH",
      error: "Set AAIS_MS_GRAPH_TOKEN or connect Microsoft 365 for syncToGraph.",
    };
  }
  const task = store.get(taskId);
  if (!task) {
    return { ok: false, reasonCode: "AAIS_TASK_NOT_FOUND", error: `No task ${taskId}` };
  }
  if (task.graphId) {
    const patch = await callGraph(
      token,
      `me/todo/lists/${encodeURIComponent(opts?.listId || "tasks")}/tasks/${encodeURIComponent(task.graphId)}`,
      "PATCH",
      { title: task.title, status: task.status === "completed" ? "completed" : "notStarted" },
      { fetchImpl: opts?.fetchImpl },
    );
    if (!patch.ok) {
      return { ok: false, reasonCode: patch.reasonCode, error: patch.error };
    }
    return { ok: true, reasonCode: "GRAPH_SYNC_TO_OK", exported: [task] };
  }
  const created = await graphCreateTodoTask(token, task.title, {
    fetchImpl: opts?.fetchImpl,
    listId: opts?.listId,
  });
  if (!created.ok) {
    return { ok: false, reasonCode: created.reasonCode, error: created.error };
  }
  const graphId = String((created.data as { id?: string })?.id || "");
  const updated = store.update(task.id, { graphId: graphId || undefined, source: "aais" });
  return {
    ok: true,
    reasonCode: "GRAPH_SYNC_TO_OK",
    exported: updated ? [updated] : [task],
  };
}

File: src/aais_tasks/graph_sync.py
"""Graph ↔ AAIS task sync (fail closed; no silent fake).

# Mythic: Graph ↔ AAIS task bridge
# Engineering: sync_from_graph / sync_to_graph
"""

from __future__ import annotations

from typing import Any

from src.aais_tasks.aais_task_store import AaisTaskStore
from src.operator_middleware_plugs.clients.graph_client import (
    call_graph,
    graph_create_todo_task,
    graph_list_todo_tasks,
)


def sync_from_graph(
    store: AaisTaskStore,
    token: str | None,
    *,
    list_id: str = "tasks",
) -> dict[str, Any]:
    if not token:
        return {
            "ok": False,
            "needs_auth": True,
            "reason_code": "GRAPH_SYNC_NEEDS_AUTH",
            "error": "Set AAIS_MS_GRAPH_TOKEN for syncFromGraph.",
        }
    listed = graph_list_todo_tasks(token, list_id=list_id)
    if not listed.get("ok"):
        return {
            "ok": False,
            "reason_code": listed.get("reason_code"),
            "error": listed.get("error"),
        }
    value = (listed.get("data") or {}).get("value") if isinstance(listed.get("data"), dict) else []
    imported = []
    for item in value or []:
        if not isinstance(item, dict):
            continue
        graph_id = str(item.get("id") or "")
        if not graph_id:
            continue
        existing = next((t for t in store.list() if t.graph_id == graph_id), None)
        status = "completed" if str(item.get("status") or "").lower() == "completed" else "notStarted"
        if existing:
            updated = store.update(
                existing.id,
                title=str(item.get("title") or existing.title),
                status=status,
                source="graph",
            )
            if updated:
                imported.append(updated.to_dict())
            continue
        task = store.create(
            title=str(item.get("title") or "Graph task"),
            status=status,
            source="graph",
            graph_id=graph_id,
        )
        imported.append(task.to_dict())
    return {"ok": True, "reason_code": "GRAPH_SYNC_FROM_OK", "imported": imported}


def sync_to_graph(
    store: AaisTaskStore,
    token: str | None,
    task_id: str,
    *,
    list_id: str = "tasks",
) -> dict[str, Any]:
    if not token:
        return {
            "ok": False,
            "needs_auth": True,
            "reason_code": "GRAPH_SYNC_NEEDS_AUTH",
            "error": "Set AAIS_MS_GRAPH_TOKEN for syncToGraph.",
        }
    task = store.get(task_id)
    if not task:
        return {"ok": False, "reason_code": "AAIS_TASK_NOT_FOUND", "error": f"No task {task_id}"}
    if task.graph_id:
        patch = call_graph(
            token,
            f"me/todo/lists/{list_id}/tasks/{task.graph_id}",
            method="PATCH",
            body={
                "title": task.title,
                "status": "completed" if task.status == "completed" else "notStarted",
            },
        )
        if not patch.get("ok"):
            return {"ok": False, "reason_code": patch.get("reason_code"), "error": patch.get("error")}
        return {"ok": True, "reason_code": "GRAPH_SYNC_TO_OK", "exported": [task.to_dict()]}
    created = graph_create_todo_task(token, task.title, list_id=list_id)
    if not created.get("ok"):
        return {"ok": False, "reason_code": created.get("reason_code"), "error": created.get("error")}
    data = created.get("data") if isinstance(created.get("data"), dict) else {}
    graph_id = str((data or {}).get("id") or "")
    updated = store.update(task.id, graph_id=graph_id or None, source="aais")
    return {
        "ok": True,
        "reason_code": "GRAPH_SYNC_TO_OK",
        "exported": [(updated or task).to_dict()],
    }
